ktor-http/ktor-http-cio/jvm/src/io/ktor/http/cio/websocket/Frame.kt

```diff
@@ -78,98 +44,100 @@ sealed class Frame(val fin: Boolean, val frameType: FrameType, val buffer: ByteB
+        data: ByteArray,
+        disposableHandle: DisposableHandle
+    ) : Frame(true, FrameType.PONG, data, disposableHandle) {
+        actual constructor(packet: ByteReadPacket) : this(packet.readBytes())
+        constructor(
+            buffer: ByteBuffer,
+            disposableHandle: DisposableHandle = NonDisposableHandle
+        ) : this(buffer.moveToByteArray(), disposableHandle)
+
+        constructor(buffer: ByteBuffer) : this(buffer.moveToByteArray(), NonDisposableHandle)
     }
 
-    override fun toString() = "Frame $frameType (fin=$fin, buffer len = $initialSize)"
+    override fun toString() = "Frame $frameType (fin=$fin, buffer len = ${data.size})"
 
     /**
      * Creates a frame copy
      */
-    fun copy() = byType(fin, frameType, ByteBuffer.allocate(buffer.remaining()).apply { buffer.slice().moveTo(this); clear() })
+    actual fun copy(): Frame = byType(fin, frameType, data.copyOf())
 
-    companion object {
-        private val Empty = ByteBuffer.allocate(0)
+    actual companion object {
+        private val Empty: ByteArray = ByteArray(0)
 
         /**
          * Create a particular [Frame] instance by frame type
          */
-        fun byType(fin: Boolean, frameType: FrameType, buffer: ByteBuffer): Frame = when (frameType) {
-            FrameType.BINARY -> Binary(fin, buffer)
-            FrameType.TEXT -> Text(fin, buffer)
-            FrameType.CLOSE -> Close(buffer)
-            FrameType.PING -> Ping(buffer)
-            FrameType.PONG -> Pong(buffer)
+        actual fun byType(
+            fin: Boolean,
+            frameType: FrameType,
+            data: ByteArray
+        ): Frame = when (frameType) {
+            FrameType.BINARY -> Binary(fin, data)
+            FrameType.TEXT -> Text(fin, data)
+            FrameType.CLOSE -> Close(data)
+            FrameType.PING -> Ping(data)
+            FrameType.PONG -> Pong(data)
         }
-    }
-}
-
-/**
- * Read text content from text frame. Shouldn't be used for fragmented frames: such frames need to be reassembled first
- */
-fun Frame.Text.readText(): String {
-    require(fin) { "Text could be only extracted from non-fragmented frame" }
-    return Charsets.UTF_8.decode(buffer.duplicate()).toString()
-}
 
-/**
- * Read binary content from a frame. For fragmented frames only returns this fragment.
- */
-fun Frame.readBytes(): ByteArray {
-    return buffer.duplicate().moveToByteArray()
+        /**
+         * Create a particular [Frame] instance by frame type
+         */
+        fun byType(fin: Boolean, frameType: FrameType, buffer: ByteBuffer): Frame =
+            byType(fin, frameType, buffer.moveToByteArray())
+    }
 }
 
 /**
  * Read close reason from close frame or null if no close reason provided
  */
 fun Frame.Close.readReason(): CloseReason? {
-    if (buffer.remaining() < 2) {
+    if (data.size < 2) {
         return null
     }
 
     buffer.mark()
-    val code = buffer.getShort()
+    val code = buffer.short
```

Comment:
`buffer.short` looks like there is some kind short buffer representation or a buffer or short integers rather that reading a short integer number. This is why I usually don't like to apply the intension.